flycheck-rust.el

```diff
@@ -94,6 +95,37 @@ Return non-nil if PROJECT-ROOT is a binary crate, nil otherwise."
   (let ((root-dir (file-name-directory project-root)))
     (file-exists-p (expand-file-name "src/main.rs" root-dir))))
 
+(defun flycheck-rust-find-target (file-name)
+  "Find and return the cargo target associated with the given file.
+
+FILE-NAME is the name of the file that is matched against the
+`src_path' value in the list `targets' returned by `cargo
+read-manifest'.  If there is no match, the first target is
+returned by default.
+
+Return a cons cell (TYPE . NAME), where TYPE is the target
+type (lib or bin), and NAME the target name (usually, the crate
+name)."
+  (let ((target-type)
+        (target-name)
+        (json-array-type 'list))
+    (let-alist (with-temp-buffer
+                 (call-process "cargo" nil t nil "read-manifest")
+                 (goto-char (point-min))
+                 (json-read))
+      ;; Assuming there is always at least one target
```

Comment:
You can simplify the entire body of this `let-alist` with:

```elisp
(let-alist ((target (seq-find (lambda (target)
                                (let-alist target (string= file-name .src_path)))
                              .targets (car .targets))))
  (cons (car .kind) .name))
```

which is not only shorter but also gets rid of the mutated variables `target-type` and `target-name`.  Pure stateless code is always easier to follow :)

Please add a `seq` dependency to `Package-Requires` at the top of this file as well.